src/theory/arith/theory_arith_private.cpp

```diff
@@ -4684,55 +4745,116 @@ bool TheoryArithPrivate::rowImplicationCanBeApplied(RowIndex ridx, bool rowUp, C
+        Debug("arith::pf::prop") << "exp len: " << explain.size() << std::endl;
         // Using the information from the above comment we assemble a conflict
         // AND in coefficient order
         NodeBuilder<> conflictInFarkasCoefficientOrder(kind::AND);
         conflictInFarkasCoefficientOrder << implication[1].negate();
         for (const Node& antecedent : implication[0])
         {
-          Debug("arith::prop") << "  ante: " << antecedent << std::endl;
+          Debug("arith::pf::prop") << "  ante: " << antecedent << std::endl;
           conflictInFarkasCoefficientOrder << antecedent;
         }
 
-        Assert(coeffs != RationalVectorPSentinel);
-        Assert(conflictInFarkasCoefficientOrder.getNumChildren()
-               == coeffs->size());
-        if (std::all_of(explain.begin(), explain.end(), [](ConstraintCP c) {
-              return c->isAssumption() || c->hasIntTightenProof();
-            }))
+        if (d_containing.d_proofRecorder
+            && coeffs->size() == clause.getNumChildren())
         {
-          d_containing.d_proofRecorder->saveFarkasCoefficients(
-              conflictInFarkasCoefficientOrder, coeffs);
+          Assert(coeffs != RationalVectorPSentinel);
+          Assert(conflictInFarkasCoefficientOrder.getNumChildren()
+                 == coeffs->size());
+          if (std::all_of(explain.begin(), explain.end(), [](ConstraintCP c) {
+                return c->isAssumption() || c->hasIntTightenProof();
+              }))
+          {
+            d_containing.d_proofRecorder->saveFarkasCoefficients(
+                conflictInFarkasCoefficientOrder, coeffs);
+          }
         }
       })
-      outputLemma(clause);
+
+      if (options::proofNew())
+      {
+        // We can prove this lemma from Farkas...
+        std::vector<std::shared_ptr<ProofNode>> conflictPfs;
+        // Assume the negated getLiteral version of the implied constaint
+        // then rewrite it into proof normal form.
+        conflictPfs.push_back(
+            d_pnm->mkNode(PfRule::MACRO_SR_PRED_TRANSFORM,
+                          d_pnm->mkAssume(implied->getLiteral().negate()),
+                          {implied->getNegation()->getProofLiteral()}));
+        // Add the explaination proofs.
+        for (const auto constraint : explain)
+        {
+          NodeBuilder<> nb;
+          conflictPfs.push_back(constraint->externalExplainByAssertions(nb));
+        }
+        // Collect the farkas coefficients, as nodes.
+        std::vector<Node> farkasCoefficients;
+        farkasCoefficients.reserve(coeffs->size());
+        auto nm = NodeManager::currentNM();
+        std::transform(
+            coeffs->begin(),
+            coeffs->end(),
+            std::back_inserter(farkasCoefficients),
+            [nm](const Rational& r) { return nm->mkConst<Rational>(r); });
+
+        // Prove bottom.
+        auto sumPf = d_pnm->mkNode(
```

Comment:
nit: Generally I prefer not using auto (IMHO I find types make things easier to read).